File: production_engine/experimental/hierarchical_rag/hierarchical_retriever.py
"""
Hierarchical Retriever: Precision child matching with parent contextual framing.
"""

import re
from typing import List, Dict, Any, Tuple
from production_engine.schemas import AssessmentTarget
from production_engine.experimental.hierarchical_rag.schemas import (
    HierarchicalEvidenceStore, ChildChunk, ParentChunk, HierarchicalRetrievedEvidence
)


class HierarchicalRetriever:
    """Retrieves target evidence by scoring child chunks and expanding to parent context windows."""

    @classmethod
    def tokenize(cls, text: str) -> List[str]:
        return [w.lower() for w in re.findall(r"\b\w{3,}\b", text)]

    @classmethod
    def retrieve_for_target(
        cls,
        store: HierarchicalEvidenceStore,
        target: AssessmentTarget,
        top_k_children: int = 2
    ) -> HierarchicalRetrievedEvidence:
        query_text = f"{target.concept_name} {target.what_taught} {target.why_assessed}"
        q_tokens = cls.tokenize(query_text)
        q_set = set(q_tokens)

        scored_children: List[Tuple[float, ChildChunk]] = []

        for child in store.children:
            c_tokens = cls.tokenize(child.text)
            if not c_tokens:
                continue

            overlap = sum(1 for t in c_tokens if t in q_set)
            jaccard = overlap / max(1, len(set(c_tokens).union(q_set)))
            
            # Boost if exact concept name appears in child text
            boost = 1.5 if target.concept_name.lower() in child.text.lower() else 1.0
            score = round(jaccard * boost, 4)

            scored_children.append((score, child))

        scored_children.sort(key=lambda x: x[0], reverse=True)
        top_children = [c for _, c in scored_children[:top_k_children]]

        if not top_children and store.children:
            top_children = store.children[:1]

        matched_child_eids = [c.evidence_id for c in top_children]
        
        # Identify and deduplicate parent contexts
        parent_eids = list(dict.fromkeys([c.parent_id for c in top_children if c.parent_id in store.parent_map]))
        
        # Build precise citation spans
        citation_spans = []
        for c in top_children:
            if c.time_span:
                s, e = c.time_span.start, c.time_span.end
                citation_spans.append(f"{c.evidence_id} ({int(s//60)}m{int(s%60):02d}s - {int(e//60)}m{int(e%60):02d}s)")
            else:
                citation_spans.append(f"{c.evidence_id}")

        # Assemble prompt context: Parent macro-narrative with explicit highlighted child focal points
        prompt_sections = []
        total_parent_words = 0
        total_child_words = sum(c.word_count for c in top_children)

        for p_eid in parent_eids:
            parent_obj = store.parent_map[p_eid]
            total_parent_words += parent_obj.word_count
            
            # Highlight child excerpts within parent context
            prompt_sections.append(
                f"=== [PARENT CONTEXT: {parent_obj.title} ({parent_obj.evidence_id})] ===\n"
                f"{parent_obj.full_text}\n"
                f"=== [END PARENT CONTEXT] ==="
            )

        combined_prompt_content = "\n\n".join(prompt_sections)
        top_score = scored_children[0][0] if scored_children else 0.0

        return HierarchicalRetrievedEvidence(
            target_id=target.target_id,
            concept_name=target.concept_name,
            matched_child_ids=matched_child_eids,
            parent_context_ids=parent_eids,
            citation_spans_text=" | ".join(citation_spans),
            retrieved_content=combined_prompt_content,
            relevance_score=top_score,
            parent_context_length_words=total_parent_words,
            child_evidence_length_words=total_child_words
        )

    @classmethod
    def retrieve_for_plan(
        cls,
        store: HierarchicalEvidenceStore,
        targets: List[AssessmentTarget]
    ) -> Dict[str, HierarchicalRetrievedEvidence]:
        return {t.target_id: cls.retrieve_for_target(store, t) for t in targets}

File: server/engine/evidence/hierarchicalRetriever.js
/**
 * server/engine/evidence/hierarchicalRetriever.js
 *
 * Faithful port of Speech_To_Text/production_engine/experimental/hierarchical_rag/hierarchical_retriever.py.
 * Retrieves target evidence by scoring child chunks and expanding to parent context windows.
 * - Precision Child Matching: Jaccard token overlap + 1.5x target concept boost
 * - Macro Narrative Context: Expands matched children into enclosing parent context windows
 */

'use strict';

function tokenize(text = '') {
  if (!text) return [];
  return (text.toLowerCase().match(/\b\w{3,}\b/g) || []);
}

class HierarchicalRetriever {
  /**
   * Retrieve hierarchical evidence for a specific assessment target.
   * @param {Object} store - HierarchicalEvidenceStore
   * @param {Object} target - Assessment target ({ concept, subtopic, instruction, supportingEvidence, targetId })
   * @param {Number} topKChildren - Number of child chunks to retrieve (default: 2)
   * @returns {Object|null} HierarchicalRetrievedEvidence
   */
  static retrieveForTarget(store, target = {}, topKChildren = 2) {
    if (!store || !Array.isArray(store.children) || store.children.length === 0) {
      return null;
    }

    const conceptName = (target.concept || target.concept_name || '').trim();
    const whatTaught = (target.instruction || target.what_taught || target.supportingEvidence || '').trim();
    const whyAssessed = (target.subtopic || target.why_assessed || '').trim();

    const queryText = `${conceptName} ${whatTaught} ${whyAssessed}`.trim();
    const qTokens = tokenize(queryText);
    const qSet = new Set(qTokens);

    if (qSet.size === 0) {
      return null;
    }

    const scoredChildren = [];

    for (const child of store.children) {
      const cTokens = tokenize(child.text);
      if (cTokens.length === 0) continue;

      const cSet = new Set(cTokens);
      let overlap = 0;
      for (const t of cTokens) {
        if (qSet.has(t)) overlap++;
      }

      const unionSize = new Set([...cTokens, ...qTokens]).size;
      const jaccard = overlap / Math.max(1, unionSize);

      // Boost if exact concept name appears in child text
      const boost = (conceptName.length > 2 && child.text.toLowerCase().includes(conceptName.toLowerCase())) ? 1.5 : 1.0;
      const score = Math.round(jaccard * boost * 10000) / 10000;

      scoredChildren.push({ score, child });
    }

    scoredChildren.sort((a, b) => b.score - a.score);

    // Filter to positive-scoring matches if available, else top candidate
    let topChildren = scoredChildren.filter(item => item.score > 0).slice(0, topKChildren).map(item => item.child);

    if (topChildren.length === 0 && store.children.length > 0) {
      // If no positive token match, take the highest scoring child
      topChildren = [scoredChildren[0].child];
    }

    const matchedChildEids = topChildren.map(c => c.evidenceId);

    // Identify unique parent contexts preserving order
    const parentEids = [];
    for (const c of topChildren) {
      if (c.parentId && store.parentMap && store.parentMap[c.parentId]) {
        if (!parentEids.includes(c.parentId)) {
          parentEids.push(c.parentId);
        }
      }
    }

    // Build precise citation spans
    const citationSpans = [];
    for (const c of topChildren) {
      if (c.timeSpan && typeof c.timeSpan.start === 'number' && typeof c.timeSpan.end === 'number') {
        const s = c.timeSpan.start;
        const e = c.timeSpan.end;
        const sMin = Math.floor(s / 60);
        const sSec = Math.floor(s % 60);
        const eMin = Math.floor(e / 60);
        const eSec = Math.floor(e % 60);
        citationSpans.push(`${c.evidenceId} (${sMin}m${String(sSec).padStart(2, '0')}s - ${eMin}m${String(eSec).padStart(2, '0')}s)`);
      } else {
        citationSpans.push(c.evidenceId);
      }
    }

    // Assemble prompt context: Parent macro-narrative with explicit context framing
    const promptSections = [];
    let totalParentWords = 0;
    let totalChildWords = 0;

    for (const c of topChildren) {
      totalChildWords += c.wordCount || 0;
    }

    for (const pEid of parentEids) {
      const parentObj = store.parentMap[pEid];
      totalParentWords += parentObj.wordCount || 0;

      promptSections.push(
        `=== [PARENT CONTEXT: ${parentObj.title} (${parentObj.evidenceId})] ===\n` +
        `${parentObj.fullText}\n` +
        `=== [END PARENT CONTEXT] ===`
      );
    }

    const combinedPromptContent = promptSections.join('\n\n');
    const topScore = scoredChildren.length > 0 ? scoredChildren[0].score : 0.0;

    return {
      targetId: target.targetId || target.target_id || 'T_UNKNOWN',
      conceptName: conceptName,
      matchedChildIds: matchedChildEids,
      parentContextIds: parentEids,
      citationSpansText: citationSpans.join(' | '),
      retrievedContent: combinedPromptContent,
      relevanceScore: topScore,
      parentContextLengthWords: totalParentWords,
      childEvidenceLengthWords: totalChildWords
    };
  }
}

module.exports = {
  HierarchicalRetriever,
  tokenize
};
